api/rest/client/methods.go

```diff
@@ -172,3 +178,106 @@ func (c *Client) GetConnectGraph() (api.ConnectGraphSerial, error) {
 	err := c.do("GET", "/health/graph", nil, &graphS)
 	return graphS, err
 }
+
+type StatusFilter struct {
+	In, Out chan api.GlobalPinInfo
+	Done    chan struct{}
+	Err     chan error
+}
+
+func NewStatusFilter() *StatusFilter {
+	return &StatusFilter{
+		In:   make(chan api.GlobalPinInfo),
+		Out:  make(chan api.GlobalPinInfo),
+		Done: make(chan struct{}),
+		Err:  make(chan error),
+	}
+}
+
+type FilterParams struct {
+	*cid.Cid
+	Local     bool
+	Target    api.TrackerStatus
+	CheckFreq time.Duration
+}
+
+func (sf *StatusFilter) Filter(ctx context.Context, fp FilterParams) {
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				sf.Err <- ctx.Err()
+				return
+			case gblPinInfo := <-sf.In:
+				ok, err := statusReached(fp.Target, gblPinInfo)
+				if err != nil {
+					sf.Err <- err
+					return
+				}
+
+				if !ok {
+					sf.Out <- gblPinInfo
+					continue
+				}
+
+				sf.Out <- gblPinInfo
+				close(sf.Done)
+				close(sf.Out)
+				return
+			}
+		}
+	}()
+}
+
+func (sf *StatusFilter) PollStatus(ctx context.Context, c *Client, fp FilterParams) {
+	ticker := time.NewTicker(fp.CheckFreq)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			sf.Err <- ctx.Err()
+			return
+		case <-ticker.C:
+			gblPinInfo, err := c.Status(fp.Cid, fp.Local)
+			if err != nil {
+				sf.Err <- err
+				return
+			}
+			sf.In <- gblPinInfo
+		case <-sf.Done:
+			close(sf.In)
+		}
+	}
+}
+
+// WaitFor is a utility function that allows for a caller to
+// wait for a paticular status for a CID. It returns a channel
+// upon which the caller can wait for the targetStatus.
+func (c *Client) WaitFor(ctx context.Context, sf *StatusFilter, fp FilterParams) {
```

Comment:
Let WaitFor consume the channels and return `(globalPinInfo, error)`, where the first is the last status received and the second is an error (either from status or context).

Reply:
Also, in that case, I think this could also work by just receiving FilterParams and creating the StatusFilter. Which in turns means that StatusFilter does not even need to be exported.